### PR TITLE
attempt to deflake ggv2 setup test

### PR DESCRIPTION
**Pull request by yuval-k**
looking at the krt snapshot from the new logs, we can see the client hasn't connected to the gateway; with that in mind, added retries to fetching xds snapshots

### CODE REVIEW

File: projects/gateway2/setup/ggv2setup_test.go [hunk 1]
```diff
@@ -240,14 +241,41 @@ func TestScenarios(t *testing.T) {
...
-				testScenario(t, ctx, client, xdsPort, fullpath)
+				testScenario(t, ctx, setupOpts.KrtDebugger, snapCache, client, xdsPort, fullpath)
 
 			})
 		}
 	}
 }
 
-func testScenario(t *testing.T, ctx context.Context, client istiokube.CLIClient, xdsPort int, f string) {
+func dumpCache(snapCache cache.SnapshotCache) iosnapshot.SnapshotResponseData {
```
Comment (sam-heilbron) [inline projects/gateway2/setup/ggv2setup_test.go:251]: I'm ok with this for now as it's an incremental improvement, but it feels like we're building functionality into our tests that could be exposed elsewhere. This is identical to https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/servers/iosnapshot/history.go#L184, so perhaps we need to consider creating a way to use that?

File: projects/gateway2/setup/ggv2setup_test.go [hunk 2]
```diff
@@ -240,14 +241,41 @@ func TestScenarios(t *testing.T) {
...
+			cacheEntries[k] = xdsSnapshot
+		}
+	}
+	return iosnapshot.SnapshotResponseData{
+		Data:  cacheEntries,
+		Error: nil,
+	}
+
+}
+func getXdsSnapshot(snapCache cache.SnapshotCache, k string) (cache cache.Snapshot, err error) {
```
Comment (sam-heilbron) [inline projects/gateway2/setup/ggv2setup_test.go:269]: Should this live in https://github.com/solo-io/gloo/tree/main/projects/gloo/pkg/xds? If getting the snapshot from the cache can panic, should this functionality also be used in our control plane directly?

File: projects/gateway2/setup/ggv2setup_test.go [hunk 3]
```diff
@@ -312,22 +348,110 @@ func testScenario(t *testing.T, ctx context.Context, client istiokube.CLIClient,
 	fmt.Println("test done")
 }
 
-func getXdsDump(t *testing.T, ctx context.Context, xdsPort int, gwname string) xdsDump {
-	conn, err := grpc.NewClient(fmt.Sprintf("localhost:%d", xdsPort), grpc.WithTransportCredentials(insecure.NewCredentials()))
+type xdsDumper struct {
```
Comment (sam-heilbron) [inline projects/gateway2/setup/ggv2setup_test.go:351]: This already existed, so I'm ok to leave it here. It seems like a pretty useful utility that others could benefit from it it lived in a more standardized place. We've been building some utilities out for envoy here: https://github.com/solo-io/gloo/tree/main/pkg/utils/envoyutils. An alternative would be in the xds package: https://github.com/solo-io/gloo/tree/main/projects/gloo/pkg/xds

**Review by sam-heilbron — COMMENTED** [review]
Looking good! Test still is failing so I assume you're still investigating. Left some minor thoughts that don't need to be acted on, just ideas for pulling custom logic out of our tests to make them reusable

File: projects/gateway2/setup/ggv2setup_test.go (hunk 2 above)
Comment (yuval-k) [reply projects/gateway2/setup/ggv2setup_test.go:271]: it is used there - see it in `history.go

File: projects/gateway2/setup/ggv2setup_test.go (hunk 3 above)
Comment (yuval-k) [reply projects/gateway2/setup/ggv2setup_test.go:353]: This code is very specialized for this use case - this is getting xds snapshot by imitating envoy pod (as this triggers the final step of translation, so we can assert on the impact of DesetinationRules). 
I would assume that other utilities will interact with a real server with envoy connected to it, and they can probably check the `/snapshots/xds` on the server instead.

**Review by sam-heilbron — APPROVED** [review]

**Pull request merged**